src/ucp/tag/offload.c

```diff
@@ -254,9 +255,20 @@ ucp_tag_offload_do_post(ucp_request_t *req)
         return UCS_ERR_NO_RESOURCE;
     }
 
-    mdi = context->tl_rscs[wiface->rsc_index].md_index;
+    mdi     = context->tl_rscs[wiface->rsc_index].md_index;
+    md_attr = &context->tl_mds[mdi].attr;
 
-    if (ucs_unlikely(length >= worker->tm.offload.zcopy_thresh)) {
+    if (!(md_attr->cap.reg_mem_types & UCS_BIT(req->recv.mem_type))) {
+        /* Can't register this buffer on the offload iface */
+        UCP_WORKER_STAT_TAG_OFFLOAD(worker, BLOCK_MEMTYPE);
+        return UCS_ERR_IO_ERROR;
+    }
```

Comment:
to be more generic, should we check  if (!req->recv.state.dt.contig.md_map) at line 290 and return ?

Reply:
why is it better?

Reply:
if it's not registered then ucp_request_memory_reg() should return error

Reply:
btw @brminich didn't ucp_request_memory_reg() return error when the md could not register memtype (e.g no GPUdirect)? whey need this check here as well?

Reply:
no, it calls `ucp_mem_rereg_mds` which does not register memory (silently), because of this check:
```     
 } else if ((md_attr->cap.flags & UCT_MD_FLAG_REG) &&
              (md_attr->cap.reg_mem_types & UCS_BIT(mem_type))) {
```

Reply:
i think we should make ucp_mem_rereg_mds() return error in such case. and ucp_mem_map_common(), which calls it, should pass only supported MDs as the input param

Reply:
we can't add such checks in every protocol which does zero copy..